teach/concepts/data-validation/strategy.md

```diff
@@ -0,0 +1,335 @@
+            "name": "firstName"
+            // ... (all other PropertyDescriptor properties)
+        }
+    ],
+    "rule": {
+        "name": "isMandatory"
+        // ... (all other PropertyDescriptor properties)
+    }
+}
+```
+
+#### Example 2: Custom Single-Property Rule Error
+
+This error is generated by a custom rule that checks a specific format.
+
+**Scenario:** The user enters an `employeeId` that doesn't follow the required pattern.
+
+-   **Invalid Data:** `employeeId: "E-1234"`
+-   **Rule Violated:** The custom `employeeIdFormat` rule.
+
+The resulting `ValidationError` would look like this:
+
+```json
+{
+    "message": "Employee ID must be in the format 'E' followed by 5 digits (e.g., E12345).",
+    "dependentProperties": [{ "name": "employeeId" }],
+    "rule": { "name": "employeeIdFormat" }
+}
+```
+
+#### Example 3: Complex Multi-Property Rule Error
+
+This error involves a rule that compares the values of two different properties.
+
+**Scenario:** The user sets an employee's `endDate` to a date that is before their `startDate`.
+
+-   **Invalid Data:** `startDate: "2025-11-10"`, `endDate: "2025-11-01"`
+-   **Rule Violated:** The custom `endDateAfterStartDate` rule.
+
+The resulting `ValidationError` would look like this:
+
+```json
+{
+    "message": "End date must be after the start date.",
+    "dependentProperties": [{ "name": "startDate" }, { "name": "endDate" }],
+    "rule": { "name": "endDateAfterStartDate" }
+}
+```
+
+#### The Complete `invalidityState` property
+
+When multiple validation rules fail on an object instance, the `invalidityState` property would contain an array of all the corresponding `ValidationError` objects.
+
+For an employee instance with all three errors from the examples above, its `invalidityState` would look like this:
+
+```json
+"invalidityState": [
+  {
+    "message": "First Name is required and cannot be empty.",
+    "dependentProperties": [{ "name": "firstName" }],
+    "rule": {"name": "isMandatory" }
+  },
+  {
+    "message": "Employee ID must be in the format 'E' followed by 5 digits (e.g., E12345).",
+    "dependentProperties": [{ "name": "employeeId" }],
+    "rule": { "name": "employeeIdFormat" }
+  },
+  {
+    "message": "End date must be after the start date.",
+    "dependentProperties": [{ "name": "startDate" }, { "name": "endDate" }],
+    "rule": { "name": "endDateAfterStartDate" }
+  }
+]
+```
+
+### 3. Smart PropertyField Components
+
+A set of UI components called **`PropertyField`** that wrap standard inputs (text fields, number inputs, etc.) to provide labels, help text, and validation error display.
+
+Each `PropertyField` components are data-aware, knowing their associated object instance (`dataInstance`) and property (`dataTypeProperty`).
+
+This lets the UI control read the `invalidityState` directly from the data object. When validation errors occur, the control automatically displays error messages and updates its appearance (e.g., showing a red border).
```

Comment:
It's the property field that drives the display of error messages and relevant visual styling, not the controls themselves

Reply:
typo